Session.js

```diff
@@ -676,17 +676,32 @@ Session.prototype = {
 	 * The value returned by the remote code. Only values that can be serialised to JSON, plus DOM elements, can be
 	 * returned.
 	 */
+
 	execute: function (script, args) {
 		// At least FirefoxDriver 2.40.0 will throw a confusing NullPointerException if args is not an array;
 		// provide a friendlier error message to users that accidentally pass a non-array
 		if (typeof args !== 'undefined' && !Array.isArray(args)) {
 			throw new Error('Arguments passed to execute must be an array');
 		}
+		var self = this;
 
-		var result = this._post('execute', {
-			script: util.toExecuteString(script),
-			args: args || []
-		}).then(lang.partial(convertToElements, this), fixExecuteError);
+		var result = this.capabilities.syncEndpoint
+			?  executeWithEndpoint('execute/sync')
+			: executeWithEndpoint('execute');
+
+		function executeWithEndpoint(endpoint) {
```

Comment:
Instead of having this embedded function, could you pick the endpoint based on the capability, and if the `execute` endpoint doesn't work, update the capability and return `self.execute(script, args)`? Something like:

```js
execute(script, args) {
  var self = this;
  var endpoint = this.capabilities.syncEndpoint ? 'execute/sync' : 'execute';
  return self._post(endpoint, ...).then(...).catch(function (error) {
    if (...) {
      self.capabilities.syncEndpoint = true;
      return self.execute(script, args);
    }
  });
}

Reply:
I used the embedded error model following what was done in closeCurrentWindow. It also makes it so that we can throw an error if something fails without repeating too much code. If I want to catch errors in the second call to execute, I would have to repeat error handling code.